streaming/format/base/shard/dual_row.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2022-2024 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Streaming shard abstract base classes."""
+
+from typing import Any, Optional
+
+from streaming.format.base.file import ShardFile
+from streaming.format.base.shard.row import RowShard
+from streaming.stream.dir_conf import StreamDirConf
+
+__all__ = ['DualRowShard']
+
+
+class DualRowShard(RowShard):
```

Comment:
Question: What SD data format would use `DualRowShard` and/or `MonoRawShard`?

Reply:
to quote PR desc part 3.6, the "what":

```
Writer
 └─ RowWriter
     ├─ DualRowWriter
     │    ├─ JSONLWriter
     │    └─ XSVWriter
     │        ├─ CSVWriter
     │        └─ TSVWriter
     └─ MonoRowWriter
         └─ MDSWriter
```

Writers write Streaming dataset directories, one shard at a time.

```
Shard
 └─ RowShard
     ├─ DualRowShard
     │   ├─ JSONLShard
     │   └─ XSVShard
     │       ├─ CSVShard
     │       └─ TSVShard
     └─ MonoRowShard
         └─ MDSShard
```

But more importantly, the "why":

I've seen github traffic patterns indicating that our previous terminology of `Joint*` and `Split*` (e.g., `SplitWriter`, `SplitReader`) was leading to people getting confused and thinking they are about dataset splits, when they are really about the file(s) that a shard's information is split over...